File: 25/python3/main.py
#!/usr/bin/env python3

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        # Helper function to check if there are at least k nodes left
        def has_k_nodes(cur, k):
            count = 0
            while cur and count < k:
                cur = cur.next
                count += 1
            return count == k
        
        # Helper function to reverse k nodes and return the new head and the next segment
        def reverse_k_nodes(head, k):
            prev = None
            cur = head
            while k > 0:
                next_node = cur.next
                cur.next = prev
                prev = cur
                cur = next_node
                k -= 1
            return prev, cur

        # Dummy node to simplify edge case handling
        dummy = ListNode(0)
        dummy.next = head
        prev_tail = dummy

        while has_k_nodes(head, k):
            # Reverse the next k nodes
            new_head, next_segment = reverse_k_nodes(head, k)
            
            # Connect the previous part with the newly reversed segment
            prev_tail.next = new_head
            
            # Move prev_tail to the end of the reversed segment (which is now 'head')
            prev_tail = head
            head = next_segment

        # Connect the remaining part of the list (if fewer than k nodes left)
        prev_tail.next = head

        return dummy.next

File: 25/cpp/main.cpp
#include <iostream>

class Solution {
public:
    // Helper function to check if there are at least k nodes left
    bool hasKNodes(ListNode* node, int k) {
        while (node && k > 0) {
            node = node->next;
            k--;
        }
        return k == 0;
    }

    // Helper function to reverse k nodes and return the new head and next segment
    ListNode* reverseKNodes(ListNode* head, int k) {
        ListNode* prev = nullptr;
        ListNode* curr = head;
        while (k > 0) {
            ListNode* next_node = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next_node;
            k--;
        }
        return prev;
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        // Dummy node to simplify edge cases and tracking
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* prev_tail = dummy;

        while (hasKNodes(head, k)) {
            // Reverse the next k nodes
            ListNode* next_segment = head;
            for (int i = 0; i < k; ++i) {
                next_segment = next_segment->next;
            }

            // Reverse the current k-group
            ListNode* new_head = reverseKNodes(head, k);

            // Connect the previous part with the newly reversed segment
            prev_tail->next = new_head;

            // Move prev_tail to the end of the reversed segment
            prev_tail = head;
            head = next_segment;
        }

        // Connect the remaining part of the list (if fewer than k nodes left)
        prev_tail->next = head;

        return dummy->next;
    }
};
